psh/wget/wget.c

```diff
@@ -0,0 +1,500 @@
+			fprintf(stderr, "wget: Unsupported protocol %*s\n", tmp - url, url);
+			return -1;
+		}
+
+		if (!isalpha(url[0])) {
+			fprintf(stderr, "wget: Illegal url\n");
+			return -1;
+		}
+
+		/* If no protocol provided, assume http */
+		bare = url;
+	}
+	else {
+		bare = url + 7;
+	}
+
+	tmp = strrchr(bare, '/');
+	if (tmp == NULL || tmp[1] == '\0') {
+		fprintf(stderr, "wget: url missing filename!\n");
+		return -1;
+	}
+
+	common.filename = tmp + 1;
+
+	tmp = strchr(bare, '/');
+	common.path = tmp + 1;
+	common.host = strndup(bare, tmp - bare);
+	if (common.host == NULL) {
+		fprintf(stderr, "wget: Out of memory!\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static uint32_t *wget_aiToAddr(struct addrinfo *ai)
+{
+	return &(((struct sockaddr_in *)ai->ai_addr)->sin_addr.s_addr);
+}
+
+
+static int wget_connect(const char *host)
+{
+	struct addrinfo *res;
+	struct addrinfo hints = { 0 };
+	char hostaddr[INET_ADDRSTRLEN];
+	int fd = -1;
+
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+
+	printf("Resolving %s... ", host);
+	fflush(stdout);
+	if (getaddrinfo(host, "80", &hints, &res) != 0) {
+		printf("Failed\n");
+		return -1;
+	}
+
+	if (inet_ntop(res->ai_family, wget_aiToAddr(res), hostaddr, sizeof(hostaddr)) == NULL) {
+		fprintf(stderr, "nslookup: Error while converting address!\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("%s\n", hostaddr);
+
+	printf("Connecting to %s|%s|:80... ", host, hostaddr);
+	fflush(stdout);
+
+	fd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+	if (fd < 0) {
+		printf("Failed\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+
+	if (connect(fd, res->ai_addr, res->ai_addrlen) < 0) {
+		close(fd);
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("Connected\n");
+
+	freeaddrinfo(res);
+
+	return fd;
+}
+
+
+static int wget_request(void)
+{
+	char buf[256];
+	int bytes = 0, i = 0, ret;
+
+	bytes += snprintf(buf, sizeof(buf), "GET /%s HTTP/1.1\r\n", common.path);
+	if (bytes >= sizeof(buf)) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	bytes += snprintf(buf + bytes, sizeof(buf) - bytes, "Host: %s\r\nUser-Agent: Wget\r\n", common.host);
+	if (bytes >= sizeof(buf)) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	bytes += snprintf(buf + bytes, sizeof(buf) - bytes, "\r\n");
+	if (bytes > sizeof(buf)) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	while (i < bytes) {
+		ret = write(common.socket, buf + i, bytes - i);
+		if (ret < 0) {
+			return -1;
+		}
+
+		i += ret;
+	}
+
+	if (common.debug) {
+		printf("--------Request--------\n%*s\n", bytes, buf);
+	}
+
+	return 0;
+}
+
+
+static int wget_file(size_t len)
+{
+	int ret;
+	size_t left = len;
+	int i = 0;
+
+	/* Save what is left in the buffer */
+	if (common.readptr < common.buf + common.len) {
+		ret = fwrite(common.readptr, 1, common.buf + common.len - common.readptr, common.outfile);
+		if (ret < 0) {
```

Comment:
`fwrite()` is type of `size_t`.
```suggestion
		size_t chunk = common.buf + common.len - common.readptr;
		ret = fwrite(common.readptr, 1, chunk, common.outfile);
		if (ret < chunk) {
```